mlflow/shap.py

```diff
@@ -32,3 +38,75 @@ def _log_matplotlib_figure(fig, out_file, artifact_path=None):
     """
     with _log_artifact_contextmanager(out_file, artifact_path) as tmp_path:
         fig.savefig(tmp_path)
+
+
+def log_explanation(predict_function, features, artifact_path=None):
+    """
+    Log the following items generated by SHAP.
+
+        - base values
+        - shap values
+        - summary bar plot
+
+    :param predict_function: A function to compute the output of a model.
+    :param features: A matrix of features to compute SHAP values with.
+    :param artifact_path: A run-relative artifact path the explanation is saved to.
+
+    :return: A URI of the logged explanation.
+
+    .. code-block:: python
+        :caption: Example
+
+        import pandas as pd
+        from sklearn.datasets import load_boston
+        from sklearn.linear_model import LinearRegression
+
+        import mlflow
+        import mlflow.shap
+
+        # prepare training data
+        dataset = load_boston()
+        X = pd.DataFrame(dataset.data[:, :4], columns=dataset.feature_names[:4])
+        y = dataset.target
+
+        # train a model
+        model = LinearRegression()
+        model.fit(X, y)
+
+        # log an explanation
+        with mlflow.start_run() as run:
+            mlflow.shap.log_explanation(model.predict, X)
+
+        # print out logged items
+        client = mlflow.tracking.MlflowClient()
+        artifacts = [x.path for x in client.list_artifacts(run.info.run_id, "shap")]
+        print(artifacts)
+
+    .. code-block:: text
+        :caption: Output
+
+        ['shap/base_values.npy', 'shap/shap_values.npy', 'shap/summary_bar_plot.png']
+
+    """
+    import matplotlib.pyplot as plt
+    import shap
+
+    artifact_path = "shap" if artifact_path is None else artifact_path
+    explainer = shap.KernelExplainer(predict_function, shap.kmeans(features, 100))
+    shap_values = explainer.shap_values(features)
+
+    shap.summary_plot(shap_values, features, plot_type="bar", show=False)
+    fig = plt.gcf()
+    _log_matplotlib_figure(fig, _SUMMARY_BAR_PLOT_FILE_NAME, artifact_path)
+    plt.close(fig)
```

Comment:
`shap.summary_plot` returns nothing so we need to use `plt.gcf` to get the active figure object.

Reply:
Noob question, since I'm not a matplotlib expert: do we want to get the current figure object, or instead create a new one? It seems it'd be bad if `mlflow.shap.log_explanation` interfered with any active figure objects the user's created thus far

Reply:
https://github.com/slundberg/shap/blob/0fe999687f74776cf171e57f0cee4e220e3db4a6/shap/plots/_beeswarm.py#L521

`shap.summary_plot` creates a new figure so it doesn't interfere with active figures.